# Hola &#x1F600;,

Soy **Hesus Garcia** – **"Soy el único Hesus que conoces (y probablemente conocerás) 🌟"** – Sí, como "Jesús", pero con una H que me hace único. Puede sonar raro, pero créeme, ¡no lo olvidarás! Como tu revisor en Triple-Ten, estoy aquí para guiarte y ayudarte a mejorar tu código. Si algo necesita un ajuste, no hay de qué preocuparse; ¡aquí estoy para hacer que tu trabajo brille con todo su potencial! ✨

Cada vez que encuentre un detalle importante en tu código, te lo señalaré para que puedas corregirlo y así te prepares para un ambiente de trabajo real, donde el líder de tu equipo actuaría de manera similar. Si en algún momento no logras solucionar el problema, te daré más detalles para ayudarte en nuestra próxima oportunidad de revisión.

Es importante que cuando encuentres un comentario, **no los muevas, no los modifiques, ni los borres**.

---

### Formato de Comentarios

Revisaré cuidadosamente cada implementación en tu notebook para asegurar que cumpla con los requisitos y te daré comentarios de acuerdo al siguiente formato:


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Éxito</b> - ¡Excelente trabajo! Esta parte está bien implementada y contribuye significativamente al análisis de datos o al proyecto. Continúa aplicando estas buenas prácticas en futuras secciones.
    
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Atención</b> ⚠️ - Este código está correcto, pero se puede optimizar. Considera implementar mejoras para que sea más eficiente y fácil de leer. Esto fortalecerá la calidad de tu proyecto.
    
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>A resolver</b> ❗ - Aquí hay un problema o error en el código que es necesario corregir para aprobar esta sección. Por favor, revisa y corrige este punto, ya que es fundamental para la validez del análisis y la precisión de los resultados.
    
</div>

---

Al final de cada revisión, recibirás un **Comentario General del Revisor** que incluirá:

- **Aspectos positivos:** Un resumen de los puntos fuertes de tu proyecto.
- **Áreas de mejora:** Sugerencias sobre aspectos donde puedes mejorar.
- **Temas adicionales para investigar:** Ideas de temas opcionales que puedes explorar por tu cuenta para desarrollar aún más tus habilidades.

Estos temas adicionales no son obligatorios en esta etapa, pero pueden serte útiles para profundizar en el futuro.

---


Esta estructura en viñetas facilita la lectura y comprensión de cada parte del comentario final.

También puedes responderme de la siguiente manera si tienes alguna duda o quieres aclarar algo específico:


<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>
    
Aquí puedes escribir tu respuesta o pregunta sobre el comentario.
    
</div>


**¡Empecemos!** &#x1F680;


# 1. Importar librerías y cargar los datos

Cargaremos los datos de las tres regiones para analizarlos. Estos archivos contienen información de los pozos como sus características (`f0`, `f1`, `f2`) y el volumen de reservas (`product`).

In [1]:
# Librerías principales
import pandas as pd
import numpy as np

# Visualización
import matplotlib.pyplot as plt
import seaborn as sns

# Machine learning
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

# Cargar los datos
data_0 = pd.read_csv('/datasets/geo_data_0.csv')
data_1 = pd.read_csv('/datasets/geo_data_1.csv')
data_2 = pd.read_csv('/datasets/geo_data_2.csv')

# Ver primeros registros de la región 0 como ejemplo
data_0.head()

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

<div class="alert alert-block alert-success">  
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>  
<b>Éxito</b> - ¡Buen inicio! Has importado correctamente las librerías necesarias y cargado los datos de forma clara y organizada. Mostrar los primeros registros ayuda a verificar la estructura de la información desde el comienzo.  
</div>


# 2. Entrenar y evaluar el modelo para la región 0

Dividiremos los datos en entrenamiento y validación, entrenaremos un modelo de regresión lineal, haremos predicciones y evaluaremos el rendimiento con RMSE.

In [2]:
# Dividir en características (X) y objetivo (y)
features_0 = data_0.drop(['product', 'id'], axis=1)
target_0 = data_0['product']

# Dividir en entrenamiento y validación (75% - 25%)
X_train_0, X_valid_0, y_train_0, y_valid_0 = train_test_split(
    features_0, target_0, test_size=0.25, random_state=42
)

# Entrenar el modelo
model_0 = LinearRegression()
model_0.fit(X_train_0, y_train_0)

# Predecir
predictions_0 = model_0.predict(X_valid_0)
# Evaluar el modelo
rmse_0 = mean_squared_error(y_valid_0, predictions_0, squared=False)
mean_pred_0 = predictions_0.mean()

print(f"RMSE región 0: {rmse_0:.2f}")
print(f"Volumen promedio predicho: {mean_pred_0:.2f}")

RMSE región 0: 37.76
Volumen promedio predicho: 92.40


<div class="alert alert-block alert-success">  
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>  
<b>Éxito</b> - Muy bien planteado el flujo de entrenamiento para la región 0. Has dividido correctamente los datos, entrenado el modelo y calculado métricas clave como el RMSE y el promedio predicho. Todo está bien estructurado y claro.  
</div>


# 3. Entrenar y evaluar el modelo para la región 1

In [3]:
# Dividir en características y objetivo
features_1 = data_1.drop(['product', 'id'], axis=1)
target_1 = data_1['product']

# Dividir en entrenamiento y validación
X_train_1, X_valid_1, y_train_1, y_valid_1 = train_test_split(
    features_1, target_1, test_size=0.25, random_state=42
)

# Entrenar modelo
model_1 = LinearRegression()
model_1.fit(X_train_1, y_train_1)

# Predecir
predictions_1 = model_1.predict(X_valid_1)

# Evaluar
rmse_1 = mean_squared_error(y_valid_1, predictions_1, squared=False)
mean_pred_1 = predictions_1.mean()

print(f"RMSE región 1: {rmse_1:.2f}")
print(f"Volumen promedio predicho: {mean_pred_1:.2f}")

RMSE región 1: 0.89
Volumen promedio predicho: 68.71


<div class="alert alert-block alert-success">  
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>  
<b>Éxito</b> - Excelente trabajo con la región 1. El proceso está replicado correctamente, y el resultado muestra un modelo muy preciso con un RMSE sorprendentemente bajo. Bien ejecutado y claramente presentado.  
</div>


# 4. Entrenar y evaluar el modelo para la región 2

In [4]:
# Dividir en características y objetivo
features_2 = data_2.drop(['product', 'id'], axis=1)
target_2 = data_2['product']

# Dividir en entrenamiento y validación
X_train_2, X_valid_2, y_train_2, y_valid_2 = train_test_split(
    features_2, target_2, test_size=0.25, random_state=42
)

# Entrenar modelo
model_2 = LinearRegression()
model_2.fit(X_train_2, y_train_2)

# Predecir
predictions_2 = model_2.predict(X_valid_2)

# Evaluar
rmse_2 = mean_squared_error(y_valid_2, predictions_2, squared=False)
mean_pred_2 = predictions_2.mean()

print(f"RMSE región 2: {rmse_2:.2f}")
print(f"Volumen promedio predicho: {mean_pred_2:.2f}")


RMSE región 2: 40.15
Volumen promedio predicho: 94.77


<div class="alert alert-block alert-success">  
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>  
<b>Éxito</b> - Muy bien completado el análisis para la región 2. Has mantenido la coherencia en el flujo de trabajo y los resultados están bien calculados y reportados. El uso del RMSE nuevamente permite una comparación efectiva entre regiones.  
</div>


# 5. Preparar datos para cálculo de ganancias

Definimos las variables clave del negocio y comparamos el umbral mínimo de producción (111.1 unidades) con el volumen medio predicho en cada región.

In [5]:
# Variables del negocio
BUDGET = 100_000_000         # 100 millones
POZOS = 200
COSTO_POR_POZO = BUDGET / POZOS
INGRESO_POR_UNIDAD = 4500    # $ por mil barriles
UMBRAL_MINIMO = COSTO_POR_POZO / INGRESO_POR_UNIDAD

print(f"Cada pozo debe producir al menos {UMBRAL_MINIMO:.1f} mil barriles para no tener pérdidas.")

# Comparar con los valores promedio de cada región
print(f"Promedio región 0: {mean_pred_0:.2f}")
print(f"Promedio región 1: {mean_pred_1:.2f}")
print(f"Promedio región 2: {mean_pred_2:.2f}")

Cada pozo debe producir al menos 111.1 mil barriles para no tener pérdidas.
Promedio región 0: 92.40
Promedio región 1: 68.71
Promedio región 2: 94.77


<div class="alert alert-block alert-success">  
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>  
<b>Éxito</b> - Has definido correctamente las variables clave del negocio y contrastado de forma clara el umbral mínimo con los promedios predichos. Muy útil este paso para orientar las decisiones estratégicas posteriores.  
</div>


# 6. Calcular ganancias esperadas con los 200 mejores pozos de cada región

In [6]:
def calcular_ganancia(predicciones, reales):
    # Crear DataFrame con predicciones y valores reales
    df = pd.DataFrame({'pred': predicciones, 'real': reales})
    
    # Seleccionar los 200 con mayor predicción
    top_200 = df.sort_values(by='pred', ascending=False).head(200)
    
    # Calcular ganancia
    ingreso_total = top_200['real'].sum() * INGRESO_POR_UNIDAD
    ganancia = ingreso_total - BUDGET
    
    return ganancia

# Calcular ganancia esperada para cada región
ganancia_0 = calcular_ganancia(predictions_0, y_valid_0)
ganancia_1 = calcular_ganancia(predictions_1, y_valid_1)
ganancia_2 = calcular_ganancia(predictions_2, y_valid_2)

print(f"Ganancia región 0: ${ganancia_0:,.0f}")
print(f"Ganancia región 1: ${ganancia_1:,.0f}")
print(f"Ganancia región 2: ${ganancia_2:,.0f}")

Ganancia región 0: $33,591,411
Ganancia región 1: $24,150,867
Ganancia región 2: $25,985,718


<div class="alert alert-block alert-success">  
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>  
<b>Éxito</b> - Excelente implementación del cálculo de ganancias. La lógica para seleccionar los mejores pozos está bien aplicada y los resultados son claros y relevantes para la toma de decisiones.  
</div>


# 7. Evaluar riesgos y ganancias con bootstrapping
Usamos bootstrapping para analizar la distribución de ganancias y el riesgo de pérdida por región.

In [7]:
def bootstrapping(predicciones, reales, n_simulaciones=1000):
    ganancias = []

    datos = pd.DataFrame({'pred': predicciones, 'real': reales})

    for _ in range(n_simulaciones):
        muestra = datos.sample(n=500, replace=True)
        top_200 = muestra.sort_values(by='pred', ascending=False).head(200)
        ganancia = top_200['real'].sum() * INGRESO_POR_UNIDAD - BUDGET
        ganancias.append(ganancia)

    ganancias = pd.Series(ganancias)
    media = ganancias.mean()
    intervalo = ganancias.quantile([0.025, 0.975])
    riesgo = (ganancias < 0).mean()

    return media, intervalo, riesgo


In [8]:
media_0, intervalo_0, riesgo_0 = bootstrapping(predictions_0, y_valid_0)
media_1, intervalo_1, riesgo_1 = bootstrapping(predictions_1, y_valid_1)
media_2, intervalo_2, riesgo_2 = bootstrapping(predictions_2, y_valid_2)

print(f"Región 0 - Ganancia media: ${media_0:,.0f}, Intervalo 95%: {intervalo_0.values}, Riesgo: {riesgo_0:.2%}")
print(f"Región 1 - Ganancia media: ${media_1:,.0f}, Intervalo 95%: {intervalo_1.values}, Riesgo: {riesgo_1:.2%}")
print(f"Región 2 - Ganancia media: ${media_2:,.0f}, Intervalo 95%: {intervalo_2.values}, Riesgo: {riesgo_2:.2%}")

Región 0 - Ganancia media: $4,154,411, Intervalo 95%: [-1014316.09480928  9369128.413349  ], Riesgo: 5.60%
Región 1 - Ganancia media: $4,327,724, Intervalo 95%: [ 473971.77979736 8159569.76327463], Riesgo: 1.40%
Región 2 - Ganancia media: $3,933,228, Intervalo 95%: [-1649977.04332844  9360676.12739465], Riesgo: 6.20%


## Conclusiones finales

Tras entrenar modelos de regresión lineal para predecir el volumen de reservas en tres regiones y evaluar las ganancias esperadas, aplicamos bootstrapping para estimar el riesgo de pérdidas.

- La región 1 presenta el **menor riesgo de pérdida (1.4%)**, inferior al umbral del 2.5% exigido.
- Su ganancia media ajustada al riesgo es de aproximadamente **4.33 millones de dólares**.
- Aunque la región 0 mostró ganancias mayores sin considerar riesgo, su probabilidad de pérdidas es más alta (5.6%).

**Por lo tanto, recomendamos invertir en la región 1** para el desarrollo de los 200 nuevos pozos, ya que ofrece el mejor balance entre rentabilidad y riesgo.

<div class="alert alert-block alert-success">  
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>  
<b>Éxito</b> - Muy bien aplicada la técnica de bootstrapping. Has analizado de forma precisa tanto la rentabilidad como el riesgo, y la recomendación final está bien fundamentada con datos sólidos y comparaciones claras entre regiones.  
</div>


## Comentario general del revisor 



<div class="alert alert-block alert-success">  
<b>Comentario del revisor</b> <a class="tocSkip"></a>  
    
¡Felicidades! Tu proyecto está <b>aprobado</b>. Has abordado con claridad y solidez todas las etapas del análisis, desde la carga y exploración de datos hasta la evaluación de ganancias y riesgos mediante bootstrapping. El enfoque aplicado demuestra un buen dominio del flujo de trabajo en ciencia de datos con orientación al negocio.

#### Puntos Positivos:

* **Procesamiento de datos:** Excelente organización en la división de datos por región y preparación para el modelado.
* **Visualización y análisis gráfico:** Aunque no incluiste visualizaciones esta vez, los resultados se presentan con claridad y precisión.
* **Ingeniería de características:** Has seleccionado correctamente las variables relevantes sin redundancias innecesarias.
* **Comparación de modelos:** Muy buen uso del RMSE y del análisis de riesgos para fundamentar decisiones estratégicas sólidas.

#### Áreas para Seguir Investigando:

* **Vectorización y optimización de código:** Explora cómo simplificar y acelerar cálculos con operaciones vectorizadas de NumPy o pandas.
* **Técnicas avanzadas de feature engineering:** Podrías investigar combinaciones no lineales o variables derivadas que mejoren el modelo.
* **Evaluación de modelos más complejos:** Comparar la regresión lineal con modelos como Random Forest o XGBoost podría enriquecer el análisis.
* **Visualización del riesgo:** Incluir histogramas o boxplots de las distribuciones del bootstrapping facilitaría aún más la interpretación.

¡Sigue avanzando con este nivel de claridad y enfoque!

</div>
